keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -818,26 +1054,166 @@ With a single `VolumeChangePolicy`, user has to fall back to `Recursive` policy
+That use case would not be possible if all mount used `-o context`:
+
+1. An privileged Pod A runs. Its volumes are mounted without any special `-o context`, privileged pods run with `spc_t` label.
+2. An unprivileged Pod B with label `c1,c2` is about to start, and it wants to use volume already mounted to Pod A.
+   Kubelet can't mount the volume with `-o context=c1,c2`, because the volume is already mounted.
+   It needs to wait for Pod A to get deleted + its volume unmounted.
+
+### Make SELinuxMount opt-in and not opt-out
+
+I.e. the old behavior is still the default.
+
+* Safer, there is no behavior change. Users that want to benefit from the new feature must opt-in.
+* Mounting with SELinux is safe for RWOP volumes, and therefore we would like to have it enabled by default.
+  The resulting `SELinuxChangePolicy` values would be:
+  * `MountReadWriteOncePod` - mount RWOP volumes with SELinux mount option, recursive relabeling for the rest. This is the default.
+  * `Recursive` - explicit opt-out.
+  * `MountOption` - mount all volumes with SELinux mount option, needs explicit opt-in.
+  * We consider such API harder to understand.
+
+### Allow opt-in (or opt-out) globally via kubelet flags
+
+Instead of `SELinuxChangePolicy` in every PodSpec, we considered having a global flag to enable or disable the feature for all Pods in a cluster.
+For example by a new kubelet flag `--enable-selinux-mount`.
+
+This is a good "plan B", if the API changes proposed here are not acceptable.
+
+
+# Appendix 1: promQL queries
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  name: selinuxmount-test
+spec:
+  groups:
+    - name: selinuxmount.rules
+      rules:
+      # Add unique_volume_id label to kube_persistentvolume_info that identifies a volume in the storage backend.
+      # PVs A and B represent the same volume in the storage backend if and only if they have the same unique_volume_id.
+      # The string is not very human friendly.
+      - record: kube_persistentvolume_info:unique_volume_id
+        expr: |
+          label_join(kube_persistentvolume_info, "unique_volume_id", "#", "gce_persistent_disk_name", "ebs_volume_id", "azure_disk_name", "fc_wwids", "fc_lun", "fc_target_wwns", "iscsi_target_portal", "iscsi_iqn", "iscsi_lun", "nfs_server", "nfs_path", "csi_driver", "csi_volume_handle", "local_path", "host_path")
+
+      # Add pod's SecurityContext to kube_pod_container_security_context
+      - record: kube_pod_container_security_context:pod_selinux
+        expr: |
+          kube_pod_security_context * on (pod, namespace, uid) group_right(pod_selinux_user, pod_selinux_role, pod_selinux_type, pod_selinux_level) kube_pod_container_security_context
+
+      # Compute correct SELinux label of a container in final_selinux_* labels and join it into final_selinux_label label that looks like "system_r:system_u:container_t:s0:c1,c92".
+      - record: kube_pod_container_security_context:final_selinux_label
```

Comment:
I wonder if multi-stage recording rules wouldn't be more efficient/practical. E.g. something like (as long as we can ensure that the results for each recording rule don't overlap)

````
- record: kube_pod_container_security_context:final_selinux_label
  expr: |
    label_replace(
      kube_pod_container_security_context:pod_selinux{privileged="true"},
      "final_selinux_level",
      "s0",
      "",
      "")
- record: kube_pod_container_security_context:final_selinux_label
  expr: |
    label_replace(
      kube_pod_container_security_context:pod_selinux{container_selinux_level!="",privileged!="true"},
      "final_selinux_level",
      "$1",
      "container_selinux_level",
      "(.+)")
- ...
````

Reply:
IIUC, and not weighing in any optimizations currently in place for this scenario by Thanos or other metrics storage solutions (since this is going into k/k), a metrics backend such as Prometheus, in and of itself, may end up [evaluating](https://github.com/prometheus/prometheus/blob/6e68046c257212e4259372dee4e97c3cc7cb0031/promql/engine.go#L1820) all split rules when the combined one is (a) queried or (b) hits its evaluation interval. I also want to point out that we may benefit from metrics that provide value on an individual level in at-least one scenario without necessarily needing to be joined, as such metrics may contribute to noise.

Reply:
I believe this can be added under the pod collector in KSM.

Reply:
I believe this can be added under the pod collector in KSM.

Edit: I just noticed that all rules are intermediaries for the last one. I believe we could add the first three rules (that don't require more than one core object collector) to KSM as IIUC despite being intermediaries, they seem to provide value on an individual level, while employing KSM's [CRS feature-set](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/metrics/extend/customresourcestate-metrics.md) to [define](https://github.com/openshift/cluster-monitoring-operator/blob/master/assets/kube-state-metrics/custom-resource-state-configmap.yaml) the last rule (in OpenShift).

Reply:
I moved the kube-state-metrics to "Alternatives" (a new controller in KCM with metrics + events is the new proposal). I will revisit this if it gets back to the Proposal.
(Thanks for the hints!)